### PR TITLE
[Polysemy] Effects Around the User Query Effect

### PR DESCRIPTION
**Pull request by mdimjasevic**
## Checklist

 - [ ] The **PR Title** explains the impact of the change.
 - [ ] The **PR description** provides context as to why the change should occur and what the code contributes to that effect. This could also be a link to a JIRA ticket or a Github issue, if there is one.
 - [ ] If this PR changes development workflow or dependencies, they have been A) automated and B) documented under docs/developer/. All efforts have been taken to minimize development setup breakage or slowdown for co-workers.
 - [ ] If HTTP endpoint paths have been added or renamed, or feature configs have changed, the **endpoint / config-flag checklist** (see Wire-employee only backend [wiki page](https://github.com/zinfra/backend-wiki/wiki/Checklists)) has been followed.
 - [ ] If a cassandra schema migration has been added, I ran **`make git-add-cassandra-schema`** to update the cassandra schema documentation.
 - [ ] **changelog.d** contains the following bits of information ([details](https://github.com/wireapp/wire-server/blob/develop/docs/developer/changelog.md)):
   - [ ] A file with the changelog entry in one or more suitable sub-sections. The sub-sections are marked by directories inside `changelog.d`.
   - [ ] If new config options introduced: added usage description under docs/reference/config-options.md
   - [ ] If new config options introduced: recommended measures to be taken by on-premise instance operators.
   - [ ] If a cassandra schema migration is backwards incompatible (see also [these docs](https://github.com/wireapp/wire-server/blob/develop/docs/developer/cassandra-interaction.md#cassandra-schema-migrations)), measures to be taken by on-premise instance operators are explained.
   - [ ] If a data migration (not schema migration) introduced: measures to be taken by on-premise instance operators.
   - [ ] If public end-points have been changed or added: does nginz need un upgrade?
   - [ ] If internal end-points have been added or changed: which services have to be deployed in a specific order?

### CODE REVIEW

File: libs/brig-types/src/Brig/Types/Common.hs [hunk 1]
```diff
@@ -111,3 +121,74 @@ instance FromJSON ExcludedPrefix where
...
+  showsPrec _ = shows . emailKeyUniq
+
+instance Eq EmailKey where
+  (EmailKey k _) == (EmailKey k' _) = k == k'
+
+-- | Turn an 'Email' into an 'EmailKey'.
+--
+-- The following transformations are performed:
+--
+--   * Both local and domain parts are forced to lowercase to make
```
Comment (stephen-smith) [inline libs/brig-types/src/Brig/Types/Common.hs:144]: While it isn't common, the local part is "domain dependent" and can be case-sensitive.  We probably shouldn't case fold it by default, though we can if we know the domain does handle the local part case-insensitively.

Case-folding by default will make for some very hard to trace issues if the domain does use case-sensitive local parts.

File: libs/brig-types/src/Brig/Types/Common.hs [hunk 2]
```diff
@@ -111,3 +121,74 @@ instance FromJSON ExcludedPrefix where
...
+--     e-mail addresses fully case-insensitive.
+--   * "+" suffixes on the local part are stripped unless the domain
```
Comment (stephen-smith) [inline libs/brig-types/src/Brig/Types/Common.hs:146]: This is also far from universal.  If we are modeling off of GMail we would also need to strip all `.` characters.

Comment (mdimjasevic) [reply libs/brig-types/src/Brig/Types/Common.hs:146]: This is not new code added by me in this PR; rather, it's old code I moved around.

File: libs/brig-types/src/Brig/Types/Common.hs (hunk 1 above)
Comment (mdimjasevic) [reply libs/brig-types/src/Brig/Types/Common.hs:144]: Just like the rest in this module, this is not new code added by me in the PR; rather, it's just being reorganised and moved from other modules. That said, I don't think this PR should be about addressing your concern and I believe you should open a ticket and bring it to the team's attention.

**Pull request closed without merging**